shared/controlflow/codeql/controlflow/ControlFlow.qll

```diff
@@ -0,0 +1,929 @@
+        upper = true and d = -1
+        or
+        upper = false and d = 1
+      )
+    )
+  }
+
+  private GuardValueExt intersectBase2(GuardValueExt v1, GuardValue v2) {
+    result = intersectBase1(v1.asBase(), v2)
+    or
+    result = intersectBase1(v2, v1.asBase())
+  }
+
+  bindingset[v1, v2]
+  pragma[inline_late]
+  private GuardValueExt intersectRange(GuardValueExt v1, GuardValue v2) {
+    exists(int low, int high | v1 = IntRange(low, high) |
+      exists(int bound, boolean upper | v2.isIntRange(bound, upper) |
+        upper = true and result = mkRange(low, high.minimum(bound))
+        or
+        upper = false and result = mkRange(low.maximum(bound), high)
+      )
+      or
+      exists(int k |
+        v2.asIntValue() = k and
+        result.asBase() = v2 and
+        low <= k and
+        k <= high
+      )
+      or
+      not v2.isIntRange(_, _) and not exists(v2.asIntValue()) and result = v1
+    )
+  }
+
+  bindingset[v1, v2]
+  pragma[inline_late]
+  private GuardValueExt intersect(GuardValueExt v1, GuardValue v2) {
+    v1 = AnyValue() and result.asBase() = v2
+    or
+    result = intersectBase2(v1, v2)
+    or
+    result = v1 and
+    v1 instanceof BaseValue and
+    not exists(intersectBase2(v1, v2))
+    or
+    result = intersectRange(v1, v2)
+  }
+
+  bindingset[v1, gv2]
+  private predicate disjointValuesExt(GuardValueExt v1, GuardValue gv2) {
+    disjointValues(v1.asBase(), gv2)
+    or
+    exists(int low, int high | v1 = IntRange(low, high) |
+      gv2.asIntValue() < low
+      or
+      high < gv2.asIntValue()
+      or
+      exists(int bound, boolean upper | gv2.isIntRange(bound, upper) |
+        upper = true and bound < low
+        or
+        upper = false and high < bound
+      )
+    )
+  }
+
+  /** An input configuration for control flow reachability. */
+  signature module ConfigSig {
+    /**
+     * Holds if the value of `def` at `node` is a source for the reachability
+     * computation.
+     */
+    predicate source(ControlFlowNode node, SsaDefinition def);
+
+    /**
+     * Holds if the value of `def` at `node` is a sink for the reachability
+     * computation.
+     */
+    predicate sink(ControlFlowNode node, SsaDefinition def);
+
+    /**
+     * Holds if the value of `gv` is a barrier for the reachability computation.
+     * That is, paths where the tracked variable can be inferred to have the
+     * value of `gv` are excluded from the reachability analysis.
+     */
+    default predicate barrierValue(GuardValue gv) { none() }
+
+    /**
+     * Holds if the edge from `bb1` to `bb2` should be excluded from the
+     * reachability analysis.
+     */
+    default predicate barrierEdge(BasicBlock bb1, BasicBlock bb2) { none() }
+
+    /**
+     * Holds if flow through uncertain SSA updates should be included.
+     */
+    default predicate uncertainFlow() { any() }
+  }
+
+  /**
+   * Constructs a control flow reachability computation.
+   */
+  module Flow<ConfigSig Config> {
```

Comment:
Should this be called `Local` or perhaps `LocalFlow` instead?

Reply:
I don't think we'll add a `Global` counterpart - so I prefer the simple `Flow` name here.